Sources/Build/MakefileSupport.swift

```diff
@@ -0,0 +1,553 @@
+            if case .library = $0.type {
+                return false
+            }
+            return true
+        }).map({ $0.ruleName })
+
+        stream <<< "\n"
+        stream <<< ".PHONY: all" <<< "\n"
+        stream <<< "all: " 
+        stream <<< allRules.sorted().joined(separator: " ")
+        stream <<< "\n"
+        stream <<< "\t@echo --done--" <<< "\n"
+    }
+
+    private func appendCleanRule(to stream: OutputByteStream) {
+        stream <<< "\n"
+        stream <<< ".PHONY: clean" <<< "\n"
+        stream <<< "clean: " <<< "\n"
+        stream <<< "\trm -rf $(BUILD_DIR)" <<< "\n"
+        stream <<< "\t@echo --cleaned--" <<< "\n"
+    }
+
+    private func appendPreamble(to stream: OutputByteStream) {
+        stream <<< """
+            # Follow POSIX standards.
+            .POSIX:
+            
+            # Disable all inference rules.
+            .SUFFIXES:
+            
+            # Compute platform specific variables.
+            ifeq ($(shell uname -s), Darwin)
+              CC := $(shell xcrun --sdk macosx --find clang)
+              SDKROOT := $(shell xcrun --sdk macosx --show-sdk-path)
+              SWIFT_EXEC := $(shell xcrun --sdk macosx --find swiftc)
+              FRAMEWORKS_DIR := $(shell xcrun --sdk macosx --show-sdk-platform-path)/Developer/Library/Frameworks
+              EXTRA_SWIFT_FLAGS := -F $(FRAMEWORKS_DIR)
+              EXTRA_CC_FLAGS := -F $(FRAMEWORKS_DIR)
+              TARGET := x86_64-apple-macosx10.10
+              TEST_LINKER_ARGS := -Xlinker -bundle
+              NUM_THREADS := $(shell sysctl -n hw.ncpu)
+              CLANG_PLATFORM_FLAGS := -fobjc-arc -arch x86_64 -mmacosx-version-min=10.10
+            else ifeq ($(shell uname -s), Linux)
+              SDKROOT := /
+              CC := $(shell which clang)
+              SWIFT_EXEC := $(shell which swiftc)
+              EXTRA_SWIFT_FLAGS :=
+              EXTRA_CC_FLAGS :=
+              TARGET := x86_64-unknown-linux
+              TEST_LINKER_ARGS := -emit-executable -Xlinker -rpath=$ORIGIN
+              NUM_THREADS := $(shell nproc --all)
+              CLANG_PLATFORM_FLAGS :=
+            endif
+
+            MAKE_FILE_PATH := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))
+            SRCROOT := $(abspath $(MAKE_FILE_PATH)/..)
+            BUILD_DIR_PATH := $(SRCROOT)/.build
+            CONFIG := debug
+            UTILS := $(MAKE_FILE_PATH)/utils.py
+            BUILD_DIR := $(BUILD_DIR_PATH)/$(TARGET)/$(CONFIG)
+            
+            ifeq ($(CONFIG), debug)
+              SWIFT_CONFIG_FLAGS := -Onone -g -enable-testing -DDEBUG
+              CLANG_CONFIG_FLAGS := -g -DDEBUG=1 -O0
+            else
+              SWIFT_CONFIG_FLAGS := -O -enable-testing
+              CLANG_CONFIG_FLAGS := -O2
+            endif
+
+            """
+    }
+
+    func createUtils(at utilsPath: AbsolutePath) throws {
+        let stream = BufferedOutputByteStream()
+        stream <<< """
+            import os
```

Comment:
Should this have a shebang line?

Reply:
Not really needed currently but don't see why not.